standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java

```diff
@@ -5240,16 +5259,38 @@ public DropPartitionsResult drop_partitions_req(
         for (Path path : archToDelete) {
           wh.deleteDir(path, true, mustPurge, needsCm);
         }
+
+        // Uses a priority queue to delete the parents of deleted directories if empty.
+        // The parent with the largest size is always processed first. It guarantees that
+        // the emptiness of a parent won't be changed once it has been processed. So duplicated
+        // processing can be avoided.
+        PriorityQueue<PathAndPartValSize> parentsToDelete = new PriorityQueue<>();
         for (PathAndPartValSize p : dirsToDelete) {
           wh.deleteDir(p.path, true, mustPurge, needsCm);
+          addParentForDel(parentsToDelete, p);
+        }
+
+        HashSet<PathAndPartValSize> processed = new HashSet<>();
+        while (!parentsToDelete.isEmpty()) {
           try {
-            deleteParentRecursive(p.path.getParent(), p.partValSize - 1, mustPurge, needsCm);
+            PathAndPartValSize p = parentsToDelete.poll();
+            if (processed.contains(p)) {
+              continue;
+            }
+            processed.add(p);
+
+            Path path = p.path;
+            if (wh.isWritable(path) && wh.isDir(path) && wh.isEmptyDir(path)) {
```

Comment:
`wh.isDir(path) && wh.isEmptyDir(path)` seems duplicated? why we need `wh.isDir` if we already have `wh.isEmptyDir`?

Reply:
wh.isEmptyDir uses listStatus that doesn't distinguish file and dir (at least for the GCS fs implementation: https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/7825ab50c839aea43f1ff587b0e2803047af99bc/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java#L997). But I agree that isEmptyDir is enough no matter the path is a file or dir.